### PR TITLE
Fix NRE that occurred when a match on a base type has no cases

### LINKED ISSUE #3484: Crash for empty match on primitive datatype
### Dafny version

3.11.0

### Code to produce this issue

```dafny
method Foo(x: int) {
  match x {
    
  }
}
```

### Command to run and resulting output

```code
dafny verify bug.dfy


Unhandled exception. System.AggregateException: One or more errors occurred. (Object reference not set to an instance of an object.)
 ---> System.NullReferenceException: Object reference not set to an instance of an object.
   at Microsoft.Dafny.MatchFlattener.FillMissingCases(Match s) in /Users/runner/work/dafny/dafny/dafny/Source/DafnyCore/CompileNestedMatch/MatchFlattener.cs:line 94
   at Microsoft.Dafny.MatchFlattener.CompileNestedMatchStmt(NestedMatchStmt nestedMatchStmt) in /Users/runner/work/dafny/dafny/dafny/Source/DafnyCore/CompileNestedMatch/MatchFlattener.cs:line 161
   at Microsoft.Dafny.MatchFlattener.<>c__DisplayClass4_0.<FlattenNode>b__0(Node node) in /Users/runner/work/dafny/dafny/dafny/Source/DafnyCore/CompileNestedMatch/MatchFlattener.cs:line 74
   at Microsoft.Dafny.Node.Visit(Func`2 beforeChildren, Action`1 afterChildren) in /Users/runner/work/dafny/dafny/dafny/Source/DafnyCore/AST/Node.cs:line 49
   at Microsoft.Dafny.MatchFlattener.FlattenNode(Node moduleDefinition) in /Users/runner/work/dafny/dafny/dafny/Source/DafnyCore/CompileNestedMatch/MatchFlattener.cs:line 62
   at Microsoft.Dafny.MatchFlattener.PostResolve(Program program) in /Users/runner/work/dafny/dafny/dafny/Source/DafnyCore/CompileNestedMatch/MatchFlattener.cs:line 51
   at Microsoft.Dafny.Resolver.ResolveProgram(Program prog) in /Users/runner/work/dafny/dafny/dafny/Source/DafnyCore/Resolver/Resolver.cs:line 728
   at Microsoft.Dafny.Main.Resolve(Program program, ErrorReporter reporter) in /Users/runner/work/dafny/dafny/dafny/Source/DafnyCore/DafnyMain.cs:line 219
   at Microsoft.Dafny.Main.ParseCheck(IList`1 files, String programName, ErrorReporter reporter, Program& program) in /Users/runner/work/dafny/dafny/dafny/Source/DafnyCore/DafnyMain.cs:line 169
   at Microsoft.Dafny.DafnyDriver.ProcessFilesAsync(IList`1 dafnyFiles, ReadOnlyCollection`1 otherFileNames, ErrorReporter reporter, Boolean lookForSnapshots, String programId) in /Users/runner/work/dafny/dafny/dafny/Source/DafnyDriver/DafnyDriver.cs:line 351
   --- End of inner exception stack trace ---
   at System.Threading.Tasks.Task.ThrowIfExceptional(Boolean includeTaskCanceledExceptions)
   at System.Threading.Tasks.Task`1.GetResultCore(Boolean waitCompletionNotification)
   at System.Threading.Tasks.Task`1.get_Result()
   at Microsoft.Dafny.DafnyDriver.ThreadMain(String[] args) in /Users/runner/work/dafny/dafny/dafny/Source/DafnyDriver/DafnyDriver.cs:line 123
   at Microsoft.Dafny.DafnyDriver.<>c__DisplayClass10_0.<Main>b__0() in /Users/runner/work/dafny/dafny/dafny/Source/DafnyDriver/DafnyDriver.cs:line 100
   at System.Threading.Thread.StartCallback()
```
```


### What happened?

No exception

### What type of operating system are you experiencing the problem on?

Mac

### PR DESCRIPTION
**Pull request by keyboardDrummer**
Fixes #3484

<small>By submitting this pull request, I confirm that my contribution is made under the terms of the [MIT license](https://github.com/dafny-lang/dafny/blob/master/LICENSE.txt).</small>

### CODE REVIEW

**Review by atomb — APPROVED** [review]
I think this looks good, as long as you think there's a good reason that the previously-asserted invariant should no longer be expected to hold.

File: Source/DafnyCore/Verifier/Translator.TrStatement.cs [hunk 1]
```diff
@@ -806,8 +806,9 @@ public partial class Translator {
         els = null;
         CurrentIdGenerator.Pop();
       }
-      Contract.Assert(ifCmd != null); // follows from the fact that s.Cases.Count + s.MissingCases.Count != 0.
```
[the remark this answers is not in the document]
Comment (keyboardDrummer) [reply Source/DafnyCore/Verifier/Translator.TrStatement.cs:809]: It holds for matches on datatypes, but matches on primitive types don't get their MissingCases filled, so if the match doesn't have cases, `cases + missingCases == 0`. @RustanLeino do you have any objections to removing this assertion?

**Comment by RustanLeino** [comment]
I'll first give four bullets to comment on the code changes, and then I'll talk about the bigger picture.

* The fix of inserting the `?` in `dtd?.ConstructorsByName` is good.
* Likewise, the removal of `Contract.Assert(ifCmd != null);`, for the reasons that @keyboardDrummer mentioned in the PR discussion, is good.
* What isn't right in this proposed fix is what happens when `ifCmd == null`. This condition can only occur if the source expression of the `match` is a non-datatype and there are no cases. This should cause the verifier to give an error.
* The test in the new file `github-issue-3484.dfy` should give an error (but it doesn't, because of the `ifCmd == null` case I mentioned).

Here's a broader description:

In Dafny, `match` implicitly asserts that only the given `case`s can occur. This cannot be determined syntactically, so it is the verifier that checks it. If the verifier finds that none of the given `case`s applies, then it could just say "error: you missed some cases", but we can be more precise than that. So, if the source expression is a datatype, then Dafny will, if effect, add `case AMissingCase(_, _, _) => assert false;` for every case that's syntactically missing. If the verifier seems to think that control can nevertheless reach this `assert false;`, then it will give an error message that says "error: you missed the `AMissingCase` case", which is more precise.

This semantics is the same for all `match`es; that is, a `match` always implicitly asserts that only the given `case`s can occur. However, if the source type is a non-datatype, then the strategy outlined above for generating a precise error message is not good. For example, if the type is `int` or `real`, it would be impossible to add a `case` for each number that's missing. Instead, the desugaring of a `NestedMatch` builds a cascading `if` statement that ends with `else { assert false; }`. Dafny does that today, _except_ when the original `match` has no cases at all (which is the bug that this PR is aiming to fix).

I think the easiest solution is to change the following situation in `MatchFlattener.cs` around line 159:

``` c#
if (compiledMatch is null) {
  // Happens only if the nested match has no cases, create a MatchStmt with no paths.
  var result = new MatchStmt(nestedMatchStmt.RangeToken, nestedMatchStmt.Source, new List<MatchCaseStmt>(), nestedMatchStmt.UsesOptionalBraces, nestedMatchStmt.Attributes);
  FillMissingCases(result);
  return result;
```

Instead of creating a `MatchStmt` with no paths, simply create an `assert false;` statement. Put an `:error` attribute on this (Dafny) `assert false;`, so the verifier will report "error: missing cases" (or "no given case applied; in fact, no cases were given and the match might be reachable" or something even better). See `MatchFlattener.cs` around line 605 for how to add the `:error` attribute to the `assert`.

Here's another possible solution: The `MatchFlattener.cs` machinery is currently called from the Resolver. There seems to be no great reason to add in the missing cases during Resolution, because only the verifier will need them. (Note, the compiler does not need them, since it will assume the code has been verified to only take one of the given cases.) So, the functionality handled by the `FillMissingCases` method in `MatchFlattener.cs` could be moved to the verifier in some way. (One would need to check that the compiler and verifier don't require match constructs to already have all their cases.)

## Something else that can be improved

A side remark. I noticed that

``` dafny
method Foo(x: int) {
  match x {
    case 3 =>
  }
}
```

gives the error

```
Error: missing case in match statement: c: int (not all possibilities for constant 'c' have been covered)
```

This is confusing. The original program does not contain any variable or constant named `c`.

File: Source/DafnyCore/CompileNestedMatch/MatchFlattener.cs [hunk 2]
```diff
@@ -132,21 +96,21 @@ public MatchFlattener(ErrorReporter reporter, FreshIdGenerator idGenerator)
...
-      // Happens only if the match has no cases, create a Match with no cases as resolved expression and let ResolveMatchExpr handle it.
-      var result = new MatchExpr(nestedMatchExpr.tok, nestedMatchExpr.Source, new List<MatchCaseExpr>(), nestedMatchExpr.UsesOptionalBraces);
-      result.Type = nestedMatchExpr.Type;
-      FillMissingCases(result);
-      return result;
-    } else if (compiledMatch.Node is Expression expression) {
+      Reporter.Error(MessageSource.Resolver, nestedMatchExpr.Tok, "match has no cases");
+      var emptyMatch = new MatchExpr(nestedMatchExpr.tok, nestedMatchExpr.Source, new List<MatchCaseExpr>(), nestedMatchExpr.UsesOptionalBraces);
+      emptyMatch.Type = nestedMatchExpr.Type;
+      return emptyMatch;
```
Comment (keyboardDrummer) [inline Source/DafnyCore/CompileNestedMatch/MatchFlattener.cs:102]: I would have preferred to use a `HavocExpression` here instead of an empty `MatchExpr`, but it doesn't exist. Created an issue: https://github.com/dafny-lang/dafny/issues/3528

**Review by RustanLeino — DISMISSED** [review]

File: Source/DafnyCore/CompileNestedMatch/MatchFlattener.cs [hunk 3]
```diff
@@ -132,21 +96,21 @@ public MatchFlattener(ErrorReporter reporter, FreshIdGenerator idGenerator)
 
     CaseBody compiledMatch = CompilePatternPaths(state, new HoleCtx(), LinkedLists.Create(nestedMatchExpr.Source), paths);
     if (compiledMatch is null) {
-      // Happens only if the match has no cases, create a Match with no cases as resolved expression and let ResolveMatchExpr handle it.
-      var result = new MatchExpr(nestedMatchExpr.tok, nestedMatchExpr.Source, new List<MatchCaseExpr>(), nestedMatchExpr.UsesOptionalBraces);
-      result.Type = nestedMatchExpr.Type;
-      FillMissingCases(result);
-      return result;
-    } else if (compiledMatch.Node is Expression expression) {
+      Reporter.Error(MessageSource.Resolver, nestedMatchExpr.Tok, "match has no cases");
```
[the remark this answers is not in the document]
Comment (keyboardDrummer) [reply Source/DafnyCore/CompileNestedMatch/MatchFlattener.cs:99]: Note that this code is about expressions not statements. I used a resolution error here because it doesn't seem right to pass an empty match expression to the downstream code, and the previous behavior also gave a resolution error in this scenario, but I'll use `assert false; var x :| true; x`

File: Test/git-issues/github-issue-3484.dfy.expect [hunk 4]
```diff
@@ -0,0 +1,5 @@
+github-issue-3484.dfy(5,2): Error: match has no cases
```
[the remark this answers is not in the document]
Comment (keyboardDrummer) [reply Test/git-issues/github-issue-3484.dfy.expect:1]: > I'm confused by this. The code above reports this as a resolution error. Yet, this test case shows what appears to be 3 verification errors. What am I reading wrong?

The resolution error is only given for match expressions, and L5 has a match statement.

[the remark this answers is not in the document]
Comment (keyboardDrummer) [reply Test/git-issues/github-issue-3484.dfy.expect:1]: > I do suggest rewording the error, since it seems to suggest that a `match` must have cases. It is only a _reachable_ `match` that must have cases. The error message could be improved, perhaps by using the line L11/L17 message and adding a parenthetic remark that somehow communicates "(in fact, this match has no given cases and yet the verifier was unable to prove that the match is unreachable)".

I think conceptually having a match without cases is like a parse error, because it only occurs when the user didn't finish writing the code they intended to compile.

Now, even if there was an actual parse error, you could still have the parser error correct for that, and then run resolution and then run verification, so that the compiler could give feedback on the part of the program that is finished.

Ideally I would let the parser require at least one case for a match, but if there are no cases let it error correct to `case _ => *` (for expressions) and `case _ => {}` for statements. The error message in the case of no cases would be "match has no cases".

I think a user's intuition is that a match without cases is not allowed, so the error message you suggest would be received as:
- "missing case in match: not all possibilities for selector of type int have been covered" -> 
  Is this about my match having no cases?
- "(in fact, this match has no given cases" -> 
  I could see that!!
- "and yet the verifier was unable to prove that the match is unreachable" -> 
  How is that relevant?

---

I will change the current error message to "Match has no cases and this is only allowed when the verifier can prove the match is unreachable." so it doesn't imply something that's incorrect.

**Review by atomb — APPROVED** [review]
From what I can tell it looks as though all of @RustanLeino's concerns have been addressed, and it would be good to get this fix available soon.
